src/marks/text.js

```diff
@@ -377,21 +397,86 @@ const defaultWidthMap = {
 // https://exploringjs.com/impatient-js/ch_strings.html#atoms-of-text
 function defaultWidth(text, start, end) {
   let sum = 0;
-  for (let i = start; i < end; ++i) {
-    sum += defaultWidthMap[text[i]] || defaultWidthMap.e;
-    const first = text.charCodeAt(i);
-    if (first >= 0xd800 && first <= 0xdbff) {
-      // high surrogate
-      const second = text.charCodeAt(i + 1);
-      if (second >= 0xdc00 && second <= 0xdfff) {
-        // low surrogate
-        ++i; // surrogate pair
-      }
-    }
+  for (let i = start; i < end; i = readCharacter(text, i)) {
+    sum += defaultWidthMap[text[i]] ?? (isPictographic(text, i) ? 120 : defaultWidthMap.e);
   }
   return sum;
 }
 
 function monospaceWidth(text, start, end) {
-  return end - start;
+  let sum = 0;
+  for (let i = start; i < end; i = readCharacter(text, i)) {
+    sum += isPictographic(text, i) ? 200 : 100;
+  }
+  return sum;
+}
+
+function overflow(input, width, widthof, textOverflow) {
+  switch (textOverflow) {
+    case "clip-end":
+      return clip(input, width, 1, widthof, "");
+    case "ellipsis-end":
+      return clip(input, width, 1, widthof, "…");
+    case "clip-start":
+      return clip(input, width, 0, widthof, "");
+    case "ellipsis-start":
+      return clip(input, width, 0, widthof, "…");
+    case "ellipsis-middle":
+      return clip(input, width, 1 / 2, widthof, "…");
+  }
+}
+
+// Clips a text to the given width, balancing head and tail in proportion to p.
+function clip(text, width, p, widthof, insert) {
+  if (insert) width -= widthof(insert, 0, insert.length);
+  text = text.trim();
+  const head = [];
+  const tail = [];
+  const lengths = [];
+  let dropped = 0;
+  let w = 0; // width consumed by selected chars
+  for (let i = 0, j; i < text.length; i = j) {
+    j = readCharacter(text, i);
+    const char = text.slice(i, j);
+    const l = widthof(text, i, j);
+    if (w < width * p) {
```

Comment:
At the point that we test `w` here, I think we want to already have added `l` to it. I.e., we want to test whether adding the next character will exceed the allowed maximum line width. I believe that currently the code allows one too many characters before clipping. And we want `w <= width * p` instead of `w < width * p`. And we don’t want to pre-emptively subtract `widthof(insert, 0, insert.length)` from `width` (because it we can fit the line in the given length, it doesn’t matter how long the insertion text is).